### PR TITLE
feat: add labels query parameter to experiments endpoint [DET-4098]

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

add labels  query parameter to experiments endpoint

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234

### CODE REVIEW

File: proto/src/determined/api/v1/experiment.proto [hunk 1]
```diff
@@ -60,12 +60,14 @@ message GetExperimentsRequest {
     int32 limit = 4;
     // Limit experiments to those that match the description.
     string description = 5;
+    // Limit experiments to those that match the provided labels.
+    repeated string labels = 6;
```
Comment (rb-determined-ai) [inline proto/src/determined/api/v1/experiment.proto:64]: This seems like an unnecessary breakage if we have already announced this API...?

File: master/internal/api_experiment.go [hunk 2]
```diff
@@ -92,6 +92,33 @@ func (a *apiServer) GetExperiments(
...
+					if hasLabel := strings.Contains(strings.ToLower(expLabel), strings.ToLower(reqLabel)); hasLabel {
+						hasAMatch = true
+						break
+					}
+				}
+				if !hasAMatch {
+					found = false
+					break
+				}
+			}
```
Comment (rb-determined-ai) [inline master/internal/api_experiment.go:115]: question: Is this a logical AND of the labels in the request, or a logical OR of the labels?

Comment (hamidzr) [reply master/internal/api_experiment.go:115]: AND. https://determined-ai.slack.com/archives/C013B86TBPC/p1599604207019900

File: proto/src/determined/api/v1/experiment.proto (hunk 1 above)
Comment (hamidzr) [reply proto/src/determined/api/v1/experiment.proto:64]: hmm, we're adding an optional parameter what would be breaking? this is requested/needed to fully utilize the pagination on the new endpoint. for the same use cases the old one had

Comment (hamidzr) [reply proto/src/determined/api/v1/experiment.proto:64]: if you're talking about the ordering of the parameters that'd only come into play for gRPC use cases which we don't have and have not announced.

Comment (rb-determined-ai) [reply proto/src/determined/api/v1/experiment.proto:64]: ok, cool

File: master/internal/api_experiment.go (hunk 2 above)
Comment (rb-determined-ai) [inline master/internal/api_experiment.go:115]: Non-blocking: would it be possible / is it worth it to use some inline functions to reduce the multi-layered flag setting and checking in this section of code?

File: master/internal/api_experiment.go [hunk 3]
```diff
@@ -92,6 +92,33 @@ func (a *apiServer) GetExperiments(
 		if len(req.Users) != 0 && !found {
 			return false
 		}
+
+		if len(req.Labels) != 0 {
+			found = true
+
+			if len(v.Labels) == 0 {
+				return false
+			}
```
Comment (rb-determined-ai) [inline master/internal/api_experiment.go:101]: nonblocking: It seems like there ought to be a way to get rid of this section, right?  Like shouldn't the `len(v.Labels) == 0` case be handled passively by whatever for loops are enforcing the logical AND?

**Review by rb-determined-ai — APPROVED** [review]

File: master/internal/api_experiment.go (hunk 2 above)
Comment (hamidzr) [reply master/internal/api_experiment.go:115]: let me look. I think also if the filtering happened through a list of filter lambdas all separately defined could make this clearer.

File: master/internal/api_experiment.go (hunk 3 above)
Comment (hamidzr) [reply master/internal/api_experiment.go:101]: yeah that bothers me too but I couldn't immediately see a way of getting rid of that without making it less readable. :thinking:

File: master/internal/api_experiment.go [hunk 4]
```diff
@@ -92,6 +117,11 @@ func (a *apiServer) GetExperiments(
 		if len(req.Users) != 0 && !found {
 			return false
 		}
+
+		if !matchesList(v.Labels, req.Labels) {
+			return false
+		}
```
Comment (rb-determined-ai) [inline master/internal/api_experiment.go:123]: ooo, I like : )

File: master/internal/api_experiment.go [hunk 5]
```diff
@@ -23,6 +23,31 @@ import (
...
+	}
+	return false
+}
+
+// matchesList checks whether srcList contains all strings provided in matchList.
+func matchesList(srcList []string, matchList []string) bool {
+	if len(matchList) != 0 && len(srcList) == 0 {
+		return false
+	}
```
Comment (rb-determined-ai) [inline master/internal/api_experiment.go:40]: Is this clause necessary?

Comment (hamidzr) [reply master/internal/api_experiment.go:40]: you're right. nope we don't need it. :rocket:

**Pull request merged**